### PR TITLE
vm: implement immutable stackitems

### LINKED ISSUE #2516: Immutable VM items and removal of DeepCopy
neo-project/neo-vm#473 and neo-project/neo#2747. More fancy VM things.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2516.

### CODE REVIEW

File: pkg/vm/vm.go [hunk 1]
```diff
@@ -1353,14 +1374,14 @@ func (v *VM) execute(ctx *Context, op opcode.Opcode, parameter []byte) (err erro
 		elems := arr.Value().([]stackitem.Item)
 		index := len(elems) - 1
 		elem := elems[index]
+		v.estack.PushItem(elem) // push item on stack firstly, to match the reference behaviour.
```
Comment (roman-khimov) [inline pkg/vm/vm.go:1377]: Not related? Is it really needed?

File: pkg/vm/stackitem/item.go [hunk 2]
```diff
@@ -1139,12 +1171,12 @@ func (i *Buffer) Len() int {
 // DeepCopy returns a new deep copy of the provided item.
 // Values of Interop items are not deeply copied.
 // It does preserve duplicates only for non-primitive types.
-func DeepCopy(item Item) Item {
+func DeepCopy(item Item, asImmutable bool) Item {
```
Comment (roman-khimov) [inline pkg/vm/stackitem/item.go:1174]: `asImmutable` doesn't make a copy (except for `Buffer`), so we better have `DeepCopy` that does a deep copy and `MakeImmutable` that makes `item` immutable. Just to make the external interface more clear.

File: pkg/vm/stackitem/item.go [hunk 3]
```diff
@@ -1155,31 +1187,38 @@ func deepCopy(item Item, seen map[Item]Item) Item {
...
+		m.MarkAsReadOnly()
 		return m
 	case *BigInteger:
 		bi := new(big.Int).Set(it.Big())
 		return (*BigInteger)(bi)
 	case *ByteArray:
 		return NewByteArray(slice.Copy(*it))
 	case *Buffer:
+		if asImmutable {
+			return NewByteArray(slice.Copy(*it)) // TODO: ported as is from C#, but is this correct?
```
Comment (roman-khimov) [inline pkg/vm/stackitem/item.go:1220]: Yes.

File: pkg/vm/vm.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/vm/vm.go:1378]: Related and needed. `item.Remove` may panic if trying to modify immutable item. If so, then C# will have on stack `elem`, and neo-go won't (without this change).

File: pkg/vm/vm.go [hunk 4]
```diff
@@ -1301,24 +1310,27 @@ func (v *VM) execute(ctx *Context, op opcode.Opcode, parameter []byte) (err erro
...
 			// NEO 2.0 doesn't error on missing key.
 			if index >= 0 {
 				elems := t.Value().([]stackitem.MapElement)
-				v.refs.Remove(elems[index].Key)
-				v.refs.Remove(elems[index].Value)
+				key := elems[index].Key
+				val := elems[index].Value
 				t.Drop(index)
+				v.refs.Remove(key)
+				v.refs.Remove(val)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/vm.go:1334]: Fixed.

File: pkg/vm/stackitem/item.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go:1174]: > `asImmutable` doesn't make a copy (except for Buffer)

But why? Insede `DeepCopy` we firstly perform item copying, and after that mark this new copy as immutable. We perform this for each stackitem.

Also, we *always* need to mark as immutable arrays, structs and maps while deepcopying irrespective to `asImmutable`.

File: pkg/vm/stackitem/item.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

File: pkg/vm/stackitem/item.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/vm/stackitem/item.go:1174]: OK, my original impression of the C# implementation was wrong.

**Pull request merged**